pkg/monitoring/manager_test.go

```diff
@@ -123,4 +125,48 @@ func TestManager(t *testing.T) {
 		require.Equal(t, len(decodedData.Feeds), len(feeds))
 		require.Equal(t, len(decodedData.Nodes), len(nodes))
 	})
+
+	t.Run("manager can manage multiple functions", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		var subs utils.Subprocesses
+		ctx, cancel := context.WithCancel(tests.Context(t))
+
+		poller := &fakePoller{ch: make(chan interface{})}
+
+		manager := NewManager(logger.Test(t), poller)
+
+		// run two managed funcs
+		var createWG sync.WaitGroup
+		var closeWG sync.WaitGroup
+		managed := func(ctx context.Context, _ RDDData) {
+			createWG.Done()
+			<-ctx.Done()
+			closeWG.Done()
+		}
+		managedNonBlocking := func(_ context.Context, _ RDDData) {
+			createWG.Done()
+			closeWG.Done()
+		}
+		subs.Go(func() {
+			manager.Run(ctx, managed, managed, managedNonBlocking)
+		})
+
+		// send RDD update to create multiple managed funcs
+		createWG.Add(3) // expect to see 3 created
+		closeWG.Add(3)  // expect to see 3 closed on restart
+		poller.ch <- RDDData{Feeds: []FeedConfig{generateFeedConfig()}}
+		createWG.Wait() // wait for created
+
+		// ensure stop + restarting works
+		createWG.Add(3)                                                 // expect 3 new created
+		closeWG.Add(3)                                                  // expect 3 closed on shutdown
```

Comment:
right before this point the closeWG should be 1 right? since one of them hasn't returned yet. so how does adding another 3 work here?

Reply:
technically `closeWG = 2` because 1 func is nonblocking and two are

but then when an new config is received, the remaining 2 are closed down and 3 more are immediately spun up

I added 3 more before the new config trigger, otherwise there could be a race where the nonblocking call would call `Done()` and trigger a panic because the wg group hasn't accounted for the new ones. so instead i opted to preallocate the count before triggering a new config

Reply:
oh ok gotcha! i see that the context is controlled by the manager so that makes sense. thanks